polaris/dataset/_base.py

```diff
@@ -0,0 +1,360 @@
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def columns(self) -> list[str]:
+        """Return all columns for the dataset"""
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def dtypes(self) -> dict[str, np.dtype]:
+        """Return the dtype for each of the columns for the dataset"""
+        raise NotImplementedError
+
+    def load_to_memory(self):
+        """
+        Load data from zarr files to memeory
+
+        Warning: Make sure the uncompressed dataset fits in-memory.
+            This method will load the **uncompressed** dataset into memory. Make
+            sure you actually have enough memory to store the dataset.
+        """
+        data = self.zarr_data
+
+        if not isinstance(data, zarr.Group):
+            raise TypeError(
+                "The dataset zarr_root is not a valid Zarr archive. "
+                "Did you call Dataset.load_to_memory() twice?"
+            )
+
+        # NOTE (cwognum): If the dataset fits in memory, the most performant is to use plain NumPy arrays.
+        # Even if we disable chunking and compression in Zarr.
+        # For more information, see https://github.com/zarr-developers/zarr-python/issues/1395
+        self._zarr_data = load_zarr_group_to_memory(data)
+
+    @abc.abstractmethod
+    def get_data(
+        self, row: str | int, col: str, adapters: dict[str, Adapter] | None = None
+    ) -> np.ndarray | Any:
+        """Since the dataset might contain pointers to external files, data retrieval is more complicated
+        than just indexing the `table` attribute. This method provides an end-point for seamlessly
+        accessing the underlying data.
+
+        Args:
+            row: The row index in the `Dataset.table` attribute
+            col: The column index in the `Dataset.table` attribute
+            adapters: The adapters to apply to the data before returning it.
+                If None, will use the default adapters specified for the dataset.
+
+        Returns:
+            A numpy array with the data at the specified indices. If the column is a pointer column,
+                the content of the referenced file is loaded to memory.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def upload_to_hub(self, access: AccessType = "private", owner: Union[HubOwner, str, None] = None):
+        """Uploads the dataset to the Polaris Hub."""
+        raise NotImplementedError
+
+    @classmethod
+    def from_json(cls, path: str):
+        """Loads a dataset from a JSON file.
+        Overrides the method from the base class to remove the caching dir from the file to load from,
+        as that should be user dependent.
+
+        Args:
+            path: Loads a dataset specification from a JSON file.
+        """
+        with fsspec.open(path, "r") as f:
+            data = json.load(f)
+        data.pop("cache_dir", None)
+        return cls.model_validate(data)
```

Comment:
**Comment**
I think `to_json` and `from_json` should be symmetrical. If one's implementation is dependent on the sub-class, then both should be abstract.